### PR TITLE
GUI: Add improved scaler selection to the Options dialog

### PR DESCRIPTION
**Pull request by ccawley2011**
This replaces the use of the Graphics Mode drop-down with two separate drop-downs for selecting the scaler and scale factor. It also moves the code for listing scalers into the GUI, which allows the backend to be simplified.

This does have a couple of limitations at the moment, namely that old configuration files aren't updated, and that the scaler options aren't shown when the OpenGL graphics manager is selected, since it doesn't implement scalers yet. The documentation also hasn't been updated for these changes yet.

This fixes [Trac #7719](https://bugs.scummvm.org/ticket/7719).

image

### CODE REVIEW

File: gui/options.cpp [hunk 1]
```diff
@@ -342,6 +347,37 @@ void OptionsDialog::build() {
...
+				for (uint scalerIndex = 0; scalerIndex < scalerPlugins.size(); scalerIndex++) {
+					if (scumm_stricmp(scalerPlugins[scalerIndex]->get<ScalerPluginObject>().getName(), scaler.c_str()) != 0)
+						continue;
+
+					_scalerPopUp->setSelectedTag(scalerIndex);
+					updateScaleFactors(scalerIndex);
+
+					if (ConfMan.hasKey("scale_factor", _domain)) {
+						int scaleFactor = ConfMan.getInt("scale_factor", _domain);
+						if (scalerPlugins[scalerIndex]->get<ScalerPluginObject>().hasFactor(scaleFactor))
```
Comment (sev-) [inline gui/options.cpp:367]: What will happen if switching to a non-existent scale? Say, from Normal 4x to SAI, which has only 2x?

**Review by sev- — COMMENTED** [review]
I like the idea. Of course, the config autoupdate must be implemented before this could be merge. Since we had the same set of scalers for over a decade, you could simply hardcode gfx_mode -> scaler+scale_factor conversion. You do it when gfx_mode is present and remove it after the upgrade.

However, I believe, that OpenGL is set as a gfx mode which should stay intact?

File: gui/options.cpp (hunk 1 above)
Comment (ccawley2011) [reply gui/options.cpp:375]: The scale factor drop-down is automatically updated whenever the selected scaler changes, which prevents invalid combinations from being selected in the GUI. If an invalid scale factor is specified via the `scummvm.ini`, the previously used factors from the backend and the individual scaler plugins will be used as fallbacks.

**Comment by sev-** [comment]
Please rebase it and we merge.

**Comment by sev-** [comment]
Thank you!

**Pull request merged**